backend/graph.py

```diff
@@ -0,0 +1,292 @@
+user.\
+"""
+
+REPHRASE_TEMPLATE = """\
+Given the following conversation and a follow up question, rephrase the follow up \
+question to be a standalone question.
+
+Chat History:
+{chat_history}
+Follow Up Input: {question}
+Standalone Question:"""
+
+
+OPENAI_MODEL_KEY = "openai_gpt_3_5_turbo"
+ANTHROPIC_MODEL_KEY = "anthropic_claude_3_haiku"
+FIREWORKS_MIXTRAL_MODEL_KEY = "fireworks_mixtral"
+GOOGLE_MODEL_KEY = "google_gemini_pro"
+COHERE_MODEL_KEY = "cohere_command"
+
+
+class AgentState(TypedDict):
+    query: str
+    documents: list[Document]
+    messages: Annotated[list[BaseMessage], add_messages]
+
+
+def get_model(model_name: str) -> LanguageModelLike:
+    gpt_3_5 = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0, streaming=True)
+    claude_3_haiku = ChatAnthropic(
+        model="claude-3-haiku-20240307",
+        temperature=0,
+        max_tokens=4096,
+        anthropic_api_key=os.environ.get("ANTHROPIC_API_KEY", "not_provided"),
+    )
+    fireworks_mixtral = ChatFireworks(
+        model="accounts/fireworks/models/mixtral-8x7b-instruct",
+        temperature=0,
+        max_tokens=16384,
+        fireworks_api_key=os.environ.get("FIREWORKS_API_KEY", "not_provided"),
+    )
+    gemini_pro = ChatGoogleGenerativeAI(
+        model="gemini-pro",
+        temperature=0,
+        max_output_tokens=16384,
+        convert_system_message_to_human=True,
+        google_api_key=os.environ.get("GOOGLE_API_KEY", "not_provided"),
+    )
+    cohere_command = ChatCohere(
+        model="command",
+        temperature=0,
+        cohere_api_key=os.environ.get("COHERE_API_KEY", "not_provided"),
+    )
+    llm: LanguageModelLike = {
+        OPENAI_MODEL_KEY: gpt_3_5,
+        ANTHROPIC_MODEL_KEY: claude_3_haiku,
+        FIREWORKS_MIXTRAL_MODEL_KEY: fireworks_mixtral,
+        GOOGLE_MODEL_KEY: gemini_pro,
+        COHERE_MODEL_KEY: cohere_command,
+    }[model_name]
+    llm = llm.with_fallbacks(
+        [gpt_3_5, claude_3_haiku, fireworks_mixtral, gemini_pro, cohere_command]
+    )
+    return llm
+
+
+def get_retriever() -> BaseRetriever:
+    weaviate_client = weaviate.Client(
+        url=WEAVIATE_URL,
+        auth_client_secret=weaviate.AuthApiKey(api_key=WEAVIATE_API_KEY),
+    )
+    weaviate_client = Weaviate(
+        client=weaviate_client,
+        index_name=WEAVIATE_DOCS_INDEX_NAME,
+        text_key="text",
+        embedding=get_embeddings_model(),
+        by_text=False,
+        attributes=["source", "title"],
+    )
+    return weaviate_client.as_retriever(search_kwargs=dict(k=6))
+
+
+def format_docs(docs: Sequence[Document]) -> str:
+    formatted_docs = []
+    for i, doc in enumerate(docs):
+        doc_string = f"<doc id='{i}'>{doc.page_content}</doc>"
+        formatted_docs.append(doc_string)
+    return "\n".join(formatted_docs)
+
+
+def retrieve_documents(state: AgentState):
+    retriever = get_retriever()
+    messages = convert_to_messages(state["messages"])
```

Comment:
ooc why's this needed? aren't these always BaseMessages

Reply:
they are coming in as dicts, so this is just deserializing them into `BaseMessage` instances

Reply:
why are they coming in as dicts?

Reply:
I am not actually sure, @nfcampos am i doing something wrong there that it's coming in as dicts or is that an expected behavior?

Reply:
Feels like these should be serialized, yeah